spring-integration-smb/src/main/java/org/springframework/integration/smb/outbound/SmbMessageHandler.java

```diff
@@ -45,4 +47,15 @@ public SmbMessageHandler(SessionFactory<SmbFile> sessionFactory) {
 		}
 	}
 
+	@SuppressWarnings("deprecation")
+	public SmbMessageHandler(SessionFactory<SmbFile> sessionFactory, FileExistsMode mode) {
+		super(new SmbRemoteFileTemplate(sessionFactory));
+		Assert.isTrue(mode == FileExistsMode.REPLACE, "Only FileExistsMode.REPLACE is supported.");
```

Comment:
Why is that?
The `mode` is not a session factory responsibility.
This is exactly what `FileTransferringMessageHandler` does for us.

What only I would like to see in this ctor is a delegation to the super one like you do now, but with this `mode`  propagation.

In addition it would be a plus to call `setReplaceFile()` if the provided mode is a `FileExistsMode.REPLACE`.

Reply:
I'm not trying to argue with you but we can't support any mode other than FileExistsMode.REPLACE, why wouldn't we want to inform the developer that only that mode is supported?

Reply:
Not sure in your conclusion. Just follow the `this.remoteFileTemplate.send(message, this.mode)` code and see how that `mode` is used.
There is a logic like in the `RemoteFileTemplate`:
```
	if (FileExistsMode.REPLACE.equals(mode)) {
			session.write(stream, tempFilePath);
		}
		else if (FileExistsMode.APPEND.equals(mode)) {
			session.append(stream, tempFilePath);
		}
```

And this has nothing to do with an SMB protocol. We just call respective `session` method and so on.

Reply:
I get that... but to fix this bug replaceFile always has to be true. There are a few ways to fix it, only one is backwards compatible and will not drive a major version change.

1. default replaceFile=true in SmbConfig
2. remove all instances of replaceFile code in SmbSession so that the existing file deletes as intended on the SMB share in the write and rename methods
3. default replaceFile=true in all constructors for SmbMessageHandler

All the code for replaceFile and even useTempFile is redundant now, it predates any of this code for FileExistsMode.

I still don't understand what use case/requirement you are trying to solution for here with this new public SmbMessageHandler(SessionFactory<SmbFile> sessionFactory, FileExistsMode mode).

Reply:
My point is to retain the `FileTransferringMessageHandler` functionality as much as possible.
And that one opens for us only if we expose a ctor based on the `FileExistsMode`.
There is nothing special to do for the `SmbSession` since it supports an `APPEND` as is even right now without any changes.